src/lib/fountain/concurrent_fountain_decoder_sink.h

```diff
@@ -35,14 +35,37 @@ class concurrent_fountain_decoder_sink
 		return _decoder.num_done();
 	}
 
+	std::vector<std::string> get_done() const
+	{
+		std::lock_guard<std::mutex> lock(_readMutex);
+		return _done;
+	}
+
+	std::vector<double> get_progress() const
+	{
+		std::lock_guard<std::mutex> lock(_readMutex);
+		return _progress;
+	}
+
+	void update_status()
+	{
+		// called under the writeMutex -- don't hold this for long!
+		std::lock_guard<std::mutex> lock(_readMutex);
+		_done = _decoder.get_done();
```

Comment:
We could retrieve these from the streams on demand instead of caching them -- but for the moment, I'm trying to keep the read-side (the const methods) as braindead simple as possible. We want `process()` to be running unopposed.